File: elixir/lib/call_coding_clis/help.ex
defmodule CallCodingClis.Help do
  alias CallCodingClis.Parser

  @canonical_runners [
    {"opencode", "oc"},
    {"claude", "cc"},
    {"kimi", "k"},
    {"codex", "c/cx"},
    {"roocode", "rc"},
    {"crush", "cr"}
  ]

  @help_text """
  ccc — call coding CLIs

  Usage:
    ccc [controls...] "<Prompt>"
    ccc --help
    ccc -h

  Slots (in order):
    runner        Select which coding CLI to use (default: oc)
                  opencode (oc), claude (cc), kimi (k), codex (c/cx), roocode (rc), crush (cr)
    +thinking     Set thinking level: +0 (off) through +4 (max)
    :provider:model  Override provider and model
    @name         Use a named preset from config; if no preset exists, treat it as an agent

  Examples:
    ccc "Fix the failing tests"
    ccc oc "Refactor auth module"
    ccc cc +2 :anthropic:claude-sonnet-4-20250514 @reviewer "Add tests"
    ccc c +4 :openai:gpt-5.4-mini @agent "Debug the parser"
    ccc k +4 "Debug the parser"
    ccc @reviewer "Audit the API boundary"
    ccc rc "Probe RooCode"
    ccc codex "Write a unit test"

  Config:
    ~/.config/ccc/config.toml  — default runner, presets, abbreviations
  """

  def runner_checklist do
    registry = Parser.runner_registry()

    lines =
      Enum.reduce(@canonical_runners, ["Runners:"], fn {name, _alias}, acc ->
        info = Map.get(registry, name)
        binary = if info, do: info.binary, else: name

        case System.find_executable(binary) do
          nil ->
            acc ++ ["  [-] #{String.pad_trailing(name, 10)} (#{binary})  not found"]

          _resolved ->
            version = get_version(binary)
            tag = if version != "", do: version, else: "found"
            acc ++ ["  [+] #{String.pad_trailing(name, 10)} (#{binary})  #{tag}"]
        end
      end)

    Enum.join(lines, "\n")
  end

  def print_help do
    IO.write(@help_text)
    IO.write("\n")
    IO.write(runner_checklist())
    IO.write("\n")
  end

  def print_usage do
    IO.write(:stderr, "usage: ccc [controls...] \"<Prompt>\"\n")
    IO.write(:stderr, runner_checklist())
    IO.write(:stderr, "\n")
  end

  defp get_version(binary) do
    try do
      {output, exit_code} = System.cmd(binary, ["--version"], stderr_to_stdout: true)
      if exit_code == 0, do: String.trim(output) |> String.split("\n") |> hd(), else: ""
    catch
      _, _ -> ""
    end
  end
end

File: ruby/lib/call_coding_clis/help.rb
# frozen_string_literal: true

require_relative "parser"

module CallCodingClis
  module Help
    CANONICAL_RUNNERS = [
      ["opencode", "oc"],
      ["claude", "cc"],
      ["kimi", "k"],
      ["codex", "c/cx"],
      ["roocode", "rc"],
      ["crush", "cr"]
    ].freeze

    HELP_TEXT = <<~HELP
      ccc — call coding CLIs

      Usage:
        ccc [controls...] "<Prompt>"
        ccc --help
        ccc -h

      Slots (in order):
        runner        Select which coding CLI to use (default: oc)
                      opencode (oc), claude (cc), kimi (k), codex (c/cx), roocode (rc), crush (cr)
        +thinking     Set thinking level: +0 (off) through +4 (max)
        :provider:model  Override provider and model
        @name         Use a named preset from config; if no preset exists, treat it as an agent

      Examples:
        ccc "Fix the failing tests"
        ccc oc "Refactor auth module"
        ccc cc +2 :anthropic:claude-sonnet-4-20250514 @reviewer "Add tests"
        ccc c +4 :openai:gpt-5.4-mini @agent "Debug the parser"
        ccc k +4 "Debug the parser"
        ccc @reviewer "Audit the API boundary"
        ccc codex "Write a unit test"

      Config:
        ~/.config/ccc/config.toml  — default runner, presets, abbreviations
    HELP

    def self.get_version(binary)
      require "open3"
      require "timeout"
      stdout, _, status = Timeout.timeout(3) do
        Open3.capture3(binary, "--version")
      end
      return "" unless status.success? && !stdout.strip.empty?
      stdout.strip.split("\n", 2).first
    rescue Errno::ENOENT, Errno::EACCES, Timeout::Error
      ""
    end

    def self.runner_checklist
      lines = ["Runners:"]
      CANONICAL_RUNNERS.each do |name, _alias|
        info = Parser::RUNNER_REGISTRY[name]
        binary = info&.binary || name
        found = system("which", binary, out: File::NULL, err: File::NULL)
        if found
          version = get_version(binary)
          tag = version.empty? ? "found" : version
          lines << "  [+] #{name.ljust(10)} (#{binary})  #{tag}"
        else
          lines << "  [-] #{name.ljust(10)} (#{binary})  not found"
        end
      end
      lines.join("\n")
    end

    def self.print_help
      $stdout.puts HELP_TEXT
      $stdout.puts
      $stdout.puts runner_checklist
    end

    def self.print_usage
      $stderr.puts 'usage: ccc [controls...] "<Prompt>"'
      $stderr.puts runner_checklist
    end
  end
end
